### DISCUSSION TITLE
computeWithDatabase not returning ever

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#2411] [Q&A] [answered]
EDITED: The following code runs and prints `1` and then `2`. If I uncomment the commented line, `1` is printed but `2` is never printed.

```dart
void onInit() async {
  var db = Provider.of<MyDatabase>(context, listen: false);
  await db.transaction(() async {
    // await (db.select(db.contact)).get();
    print(1);
    await db.computeWithDatabase(
      computation: (dbc) async {
        await (dbc.select(dbc.contact)).get();
      },
      connect: (connection) {
        return MyDatabase(connection);
      },
    );
    print(2);
    return;
  });
}
```

If I remove the transaction the code works fine. So if inside a `transaction`, I run a query which accesses the `db` and THEN do something inside `computeWithDatabase` it gets stuck.

Note, if I don't access `dbc` inside of the `computeWithDatabase`, say I just put `await Future.delayed(Duration(seconds: 1));` the issue doesn't happen, so it has to do with accessing the db inside of the `computeWithDatabase` while previously also accessed inside the `transaction`

**Comment by simolus3** [comment, marked as answer]
Very good catch, thanks for the report!

The problem is that `computeWithDatabase` "shares" the top-level connection (typically the `NativeDatabase`) used to open the original `MyDatabase` instance. Calling `transaction` will make drift suspend the top-level connection until the transaction is finished (which is necessary because we can only use a single connection, and statements outside of transactions should not see data in uncommitted transactions). So this is a classic deadlock: The transaction is waiting on `computeWithDatabase`, and the select on the other isolate is blocked on the transaction.

I've fixed this in 3ee1c1ccd4465e44165c787ca6247d88d21ee7c6, which will make us share the transaction with the new isolate instead.

**Reply by BananaMasterz** [reply]
That's great! I was troubled for several hours till I figured the specific scenario that was causing the issue. (I guess I should have opened an issue instead of a discussion after all haha)

Glad it's fixed, thanks!
